### ISSUE TITLE
archives not working with carbondate?

### ISSUE THREAD

**Issue by phonedude** [oduwsdl/CarbonDate#25]
@ibnesayeed @grantat 

I noticed this the other day and I wonder if it has something to do with the extra memgator traffic that we saw.  carbondate no longer reports the results from archives:

$ curl -i http://carbondate.cs.odu.edu/cd/http://www.cs.odu.edu/~mln/pubs/all.html
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Content-Length: 609
Content-Type: application/json
Date: Wed, 01 Aug 2018 15:14:45 GMT
Etag: "3d4b6df46dda5b1ddbf0f34a1350a1a4a8414889"
Server: TornadoServer/4.5.2

{
  "self": "http://carbondate.cs.odu.edu/cd/http://www.cs.odu.edu/~mln/pubs/all.html",
  "uri": "http://www.cs.odu.edu/~mln/pubs/all.html",
  "estimated-creation-date": "2010-09-13T16:25:51",
  "earliest-sources": [
    "bitly.com"
  ],
  "sources": {
    "backlinks": {
      "earliest": ""
    },
    "bing.com": {
      "earliest": ""
    },
    "bitly.com": {
      "earliest": "2010-09-13T16:25:51"
    },
    "google.com": {
      "earliest": ""
    },
    "last-modified": {
      "earliest": ""
    },
    "pubdate": {
      "earliest": ""
    },
    "twitter.com": {
      "earliest": ""
    }
  }
}

and that page is in the archives:

$ curl -i http://memgator.cs.odu.edu/timemap/json/http://www.cs.odu.edu/~mln/pubs/all.html
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Access-Control-Expose-Headers: Link, Location, X-Memento-Count, X-Generator
Content-Type: application/json
Date: Wed, 01 Aug 2018 15:18:05 GMT
X-Generator: MemGator:1.0-rc7
X-Memento-Count: 34
Transfer-Encoding: chunked

{
  "original_uri": "http://www.cs.odu.edu/~mln/pubs/all.html",
  "self": "https://memgator.cs.odu.edu/timemap/json/http://www.cs.odu.edu/~mln/pubs/all.html",
  "mementos": {
    "list": [
      {
        "datetime": "2002-10-05T11:28:43Z",
        "uri": "http://web.archive.org/web/20021005112843/http://www.cs.odu.edu:80/~mln/pubs/all.html"
      },
...[deletia]...
      {
        "datetime": "2018-05-21T15:29:09Z",
        "uri": "http://web.archive.org/web/20180521152909/http://www.cs.odu.edu/~mln/pubs/all.html"
      }
    ],
    "first": {
      "datetime": "2002-10-05T11:28:43Z",
      "uri": "http://web.archive.org/web/20021005112843/http://www.cs.odu.edu:80/~mln/pubs/all.html"
    },
    "last": {
      "datetime": "2018-05-21T15:29:09Z",
      "uri": "http://web.archive.org/web/20180521152909/http://www.cs.odu.edu/~mln/pubs/all.html"
    }
  },
  "timemap_uri": {
    "link_format": "https://memgator.cs.odu.edu/timemap/link/http://www.cs.odu.edu/~mln/pubs/all.html",
    "json_format": "https://memgator.cs.odu.edu/timemap/json/http://www.cs.odu.edu/~mln/pubs/all.html",
    "cdxj_format": "https://memgator.cs.odu.edu/timemap/cdxj/http://www.cs.odu.edu/~mln/pubs/all.html"
  },
  "timegate_uri": "https://memgator.cs.odu.edu/timegate/http://www.cs.odu.edu/~mln/pubs/all.html"
}

I've noticed this with other pages too.  

$ curl -i http://carbondate.cs.odu.edu/cd/http://www.cnn.com/
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Content-Length: 605
Content-Type: application/json
Date: Wed, 01 Aug 2018 15:27:11 GMT
Etag: "fc6932434d7c6d31675d110f209d23683a4625ac"
Server: TornadoServer/4.5.2

{
  "self": "http://carbondate.cs.odu.edu/cd/http://www.cnn.com/",
  "uri": "http://www.cnn.com/",
  "estimated-creation-date": "2008-10-13T19:46:46",
  "earliest-sources": [
    "bitly.com"
  ],
  "sources": {
    "backlinks": {
      "earliest": ""
    },
    "bing.com": {
      "earliest": ""
    },
    "bitly.com": {
      "earliest": "2008-10-13T19:46:46"
    },
    "google.com": {
      "earliest": ""
    },
    "last-modified": {
      "earliest": ""
    },
    "pubdate": {
      "earliest": "2014-02-19T19:15:05"
    },
    "twitter.com": {
      "earliest": "2018-07-22T23:57:18"
    }
  }
}

**Comment by grantat** [comment]
Daily build still seems to be hitting it and working fine: https://travis-ci.org/grantat/CarbonDate. 
I do notice that it is taking very long to retrieve a CD for that URI. I'll look into it.

Looking back at the example URI-Rs they provided concerning memgator traffic, I found none of them to be present in the logs.

**Comment by grantat** [comment]
Ran both the local and server docker versions and got the expected output: 

```
{
  "self": "",
  "uri": "http://www.cs.odu.edu/~mln/pubs/all.html",
  "estimated-creation-date": "2002-09-16T09:47:38",
  "earliest-sources": [
    "web.archive.org"
  ],
  "sources": {
    "web.archive.org": {
      "uri-m": "http://web.archive.org/web/20021005112843/http://www.cs.odu.edu:80/~mln/pubs/all.html",
      "memento-datetime": "2002-09-16T09:47:38",
      "memento-pubdate": "",
      "earliest": "2002-09-16T09:47:38"
    },
    "twitter.com": {
      "earliest": "2011-06-28T09:52:55"
    },
    "bitly.com": {
      "earliest": ""
    },
    "backlinks": {
      "earliest": ""
    },
    "bing.com": {
      "earliest": ""
    },
    "last-modified": {
      "earliest": ""
    },
    "google.com": {
      "earliest": ""
    },
    "pubdate": {
      "earliest": ""
    }
  }
}
```

Which tells me that the limited resources on our server, i.e. the amount of threads being used for this application, are causing a timeout when waiting for them to complete shown in: https://github.com/oduwsdl/CarbonDate/blob/master/core.py#L125

One solution could be to implement a rate limit argument when starting the app and prevent users from repeatedly accessing the API.

**Comment by phonedude** [comment]
can we turn off other modules, like backlinks, that we know take a long time?  the archives module should work since it's the most likely to give the right answer for most URLs.

**Comment by ibnesayeed** [comment]
> Which tells me that the limited resources on our server, i.e. the amount of threads being used for this application, are causing a timeout when waiting for them to complete shown in https://github.com/oduwsdl/CarbonDate/blob/master/core.py#L125

I doubt number of threads being the issue as these threads are network I/O heavy, not processor heavy. Depending on the application server being used you may or may not have concurrent serving. The machine where we host our Docker services is a powerful machine with sufficient resources for our current load.

@grantat if you have figured the issue out and want to redeploy the service with changed configurations, I can give you access to the deployment machine or update it on your behalf. Talk to me on Slack about it. I am in SF and attending #DWebSummit, so I can't investigate it in details right now.

**Comment by ibnesayeed** [comment]
By the way, we have some performance issues with threads in IPWB which might have something to do with the library we use to connect to the IPFS. I wrote a well-documented [benchmarking script](https://github.com/ibnesayeed/ipfsapi-concurrency-test) for that a couple weeks ago over the weekend. In the process I found out that Python 3 has implemented [asyncio](https://docs.python.org/3/library/asyncio.html) which you might want to try here instead of traditional threads for concurrency.

**Comment by phonedude** [comment]
@ibnesayeed: could you please document the inter-container access problem you showed me, as well as the switch from fqdn -> relative hostname to temporarily fix the problem.

**Comment by ibnesayeed** [comment]
The issue was identified and fixed the other day. Someone added some mysterious firewall rules (cough IPTables) that was blocking requests between the docker network interface and the host machine. I reported this on Slack after it was fixed. The temporary fix was then removed and the application was restored to its previous state.

The issue did surface a problem in the code that the TimeMap endpoint is currently not configurable at run time, hence needs code change if one is willing to use a different archive or aggregator.
